freebsd/hid.c

```diff
@@ -0,0 +1,1161 @@
+		errno = EINVAL;
+		register_device_error(dev, "Zero buffer/length");
+		return -1;
+	}
+
+	struct hid_device_info *info = hid_get_device_info(dev);
+	if (!info) {
+		// hid_get_device_info will have set an error already
+		return -1;
+	}
+
+	if (info->manufacturer_string) {
+		wcsncpy(string, info->manufacturer_string, maxlen);
+		string[maxlen - 1] = L'\0';
+	}
+	else {
+		string[0] = L'\0';
+	}
+
+	return 0;
+}
+
+int HID_API_EXPORT_CALL hid_get_product_string(hid_device *dev, wchar_t *string, size_t maxlen)
+{
+	if (!string || !maxlen) {
+		errno = EINVAL;
+		register_device_error(dev, "Zero buffer/length");
+		return -1;
+	}
+
+	struct hid_device_info *info = hid_get_device_info(dev);
+	if (!info) {
+		// hid_get_device_info will have set an error already
+		return -1;
+	}
+
+	if (info->product_string) {
+		wcsncpy(string, info->product_string, maxlen);
+		string[maxlen - 1] = L'\0';
+	}
+	else {
+		string[0] = L'\0';
+	}
+
+	return 0;
+}
+
+int HID_API_EXPORT_CALL hid_get_serial_number_string(hid_device *dev, wchar_t *string, size_t maxlen)
+{
+	if (!string || !maxlen) {
+		errno = EINVAL;
+		register_device_error(dev, "Zero buffer/length");
+		return -1;
+	}
+
+	struct hid_device_info *info = hid_get_device_info(dev);
+	if (!info) {
+		// hid_get_device_info will have set an error already
+		return -1;
+	}
+
+	if (info->serial_number) {
+		wcsncpy(string, info->serial_number, maxlen);
+		string[maxlen - 1] = L'\0';
+	}
+	else {
+		string[0] = L'\0';
+	}
+
+	return 0;
+}
+
+
+HID_API_EXPORT struct hid_device_info *HID_API_CALL hid_get_device_info(hid_device *dev) {
+	int idx;
+	char path_buffer[1024];
+
+	if (dev->device_info) {
+		register_device_error(dev, NULL);
+	}
+	else {
+		// Lazy initialize device_info
+		realpath(dev->device_path, path_buffer);
+		sscanf(path_buffer, "/dev/hidraw%d", &idx);
+		struct hid_device_info *devinfo =
+			hid_create_device_info_by_hidraw_idx(dev->device_handle, idx);
+		if (devinfo != NULL) {
+			dev->device_info = devinfo->next;
+			devinfo->next = NULL;
+		} else {
+			register_device_error(dev, "Failed to create device: maybe the permission is not correct?");
+		}
+	}
+
+	// hid_create_device_info_by_hidraw_idx will set an error if needed
+	return dev->device_info;
+}
+
+int HID_API_EXPORT_CALL hid_get_indexed_string(hid_device *dev, int string_index, wchar_t *string, size_t maxlen)
+{
+	(void)string_index;
+	(void)string;
+	(void)maxlen;
+
+	errno = ENOSYS;
+	register_device_error(dev, "hid_get_indexed_string: not supported by hidraw");
+
+	return -1;
+}
+
+int HID_API_EXPORT_CALL hid_get_report_descriptor(hid_device *dev, unsigned char *buf, size_t buf_size)
+{
+	int error;
+	struct hidraw_gen_descriptor desc;
+
+	if (!buf || !buf_size) {
+		errno = EINVAL;
+		register_device_error(dev, "Zero buffer/length");
+		return -1;
+	}
+
+	register_device_error(dev, NULL);
+
+	desc.hgd_maxlen = buf_size;
+	desc.hgd_data = (void *)buf;
+	error = ioctl(dev->device_handle, HIDRAW_GET_REPORT_DESC, &desc);
```

Comment:
thans @wulf7 for the hint
We do not want to use `HIDRAW_GET_REPORT_DESC` here, since it [switches operation mode](https://github.com/freebsd/freebsd-src/blob/d8cf519492523f6dd239ab153029db62749afa52/sys/dev/hid/hidraw.c#L643) to `uhid`.
Use `HIDIOCGRDESC` instead.